cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -405,6 +421,8 @@ public void deleteApp(HttpRequest request, HttpResponder responder,
                         @PathParam("namespace-id") String namespaceId,
                         @PathParam("app-id") final String appId) throws Exception {
     ApplicationId id = validateApplicationId(namespaceId, appId);
+    String latestAppVersion = getLatestAppVersion(id);
+    id = validateApplicationVersionId(namespaceId, appId, latestAppVersion);
     applicationLifecycleService.removeApplication(id);
```

Comment:
Why do you need to know the exact version? According to the doc, it should remove all app versions. What would happen if app does not exists (what error is now and if this error stays the same)?

Reply:
so we want to get latest here - to ensure we have a valid application - since latest always returns an app version if the app exists. If we don't do this - then this api uses default "-SNAPSHOT" and references the app using -SNAPSHOT version (which is possible that it might not exist)

Reply:
I don't think you need to validate since this is removing all versions, so you don't care about if it is -SNAPSHOT or latest. You only need to check the app itself exists or not, if not, throw a 404

Reply:
+1. You are removing all versions. You should not need to retrieve the latest one

Reply:
`removeApplication` runs a few checks on appId (access check etc https://github.com/cdapio/cdap/pull/14627/files#diff-bd61e1bd8f120c6108e9b2346f1b32ab44bfc6d476617641ba2e9458166d08b4R900) - hence we need a valid appId

Reply:
That's a problem with removeApplication then.

Reply:
do you mean get the logic of getting the valid app - (latest app) should be there?

Reply:
Since it removes all versions it should not care about input version. Actually it's better if it would take a versionless identifier (or it's parts)

Reply:
so I think we should have `removeApplication()` that calls the latest app-version and runs those checks specifically?

Reply:
It should not need to retrieve the latest app. Even for the running apps check - it should check irrespective of version. What if you started an app, then edited it, then tried to delete? Old one is still running, but latest versions is not running.

Reply:
but we still need the latest app-spec here https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java#L1077 so I think we should get it in the deleteApp function function?

Reply:
Well, if you need it in certain place, you can retrieve it. But need to understand why as it may indicate a problem.
Generally, I see that we need some kind of application id object without version as a lot of things are referring application (e.g. triggers or active schedules) and not specific version of application and using ApplicationId all over the place even when version is not needed is very error-prone.

BTW: `getAllAppVersions` call is very dangerous and must be refactored in a streaming manner. I would not be surprised if we see applications with 100+ or even 1000+ versions for really creative people. Filled https://cdap.atlassian.net/browse/CDAP-19783